src/kirby/directives/fit-heading/fit-heading.directive.ts

```diff
@@ -0,0 +1,150 @@
+      fontSize: this.getFontSize('l'),
+      lineHeight: this.getLineHeight('l'),
+    },
+    {
+      name: 'h3',
+      fontSize: this.getFontSize('m'),
+      lineHeight: this.getLineHeight('m'),
+    },
+  ];
+
+  constructor(
+    private elementRef: ElementRef,
+    private renderer: Renderer2,
+    private resizeObserverService: ResizeObserverService
+  ) {}
+
+  ngOnInit(): void {
+    if (this.config && this.config.maxLines) {
+      this.observeResize();
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.resizeObserverService.unobserve(this.elementRef);
+    this.renderer.removeChild(this.elementRef.nativeElement, this.clone);
+  }
+
+  private getFontSize(size): string {
+    const fontSize = this.getScssValue(['global', '$font-sizes', 'value', size]);
+    if (fontSize) {
+      return `${fontSize.value}${fontSize.unit}`;
+    }
+  }
+
+  private getLineHeight(size): string {
+    const lineHeight = this.getScssValue(['global', '$line-height', 'value', size]);
+    if (lineHeight) {
+      return `${lineHeight.value}${lineHeight.unit}`;
+    }
+  }
+
+  private getScssValue(path: string[]) {
+    let node = this.scssVariables;
+    for (let step of path) {
+      node = node[step];
+      if (!node) {
+        break;
+      }
+    }
+    return node;
+  }
+
+  private observeResize(): void {
+    this.resizeObserverService.observe(this.elementRef, (entry: ResizeObserverEntry) => {
+      this.handleResize(entry);
+    });
+  }
+
+  private handleResize(entry: ResizeObserverEntry) {
+    if (!this.shouldScale(entry.target)) return;
+
+    // Set width to determine at next resize if header should be scaled up again
+    this.previousWidth = entry.target.clientWidth;
+    this.scaleHeader();
+  }
+
+  private shouldScale(el: Element): boolean {
+    const height = el.clientHeight;
+
+    if (height === 0 || this.isScalingHeader) return false;
+
+    const lineHeight = parseInt(
+      window.getComputedStyle(this.elementRef.nativeElement).getPropertyValue('line-height')
+    );
+
+    const lines = height / lineHeight;
+    return lines > this.config.maxLines || this.previousWidth < el.clientWidth;
+  }
+
+  private scaleHeader(): void {
+    this.isScalingHeader = true;
+
+    if (!this.clone) {
+      this.clone = this.generateClone();
+      this.renderer.appendChild(this.elementRef.nativeElement, this.clone);
+    }
+
+    this.renderer.setStyle(this.clone, 'width', `${this.elementRef.nativeElement.clientWidth}px`);
+
+    const fallbackSize = this.headingSizes[this.headingSizes.length - 1];
+    const fittedSize = this.headingSizes.find(this.canFitHeading.bind(this)) || fallbackSize;
+
+    this.setSize(this.elementRef.nativeElement, fittedSize);
+    this.isScalingHeader = false;
+  }
+
+  private canFitHeading(size: HeadingSize) {
+    this.setSize(this.clone, size);
+    const lines = this.clone.clientHeight / parseInt(size.lineHeight);
+    return lines <= this.config.maxLines;
+  }
+
+  private generateClone(): Element {
```

Comment:
Maybe call this generateHostElementClone?

Reply:
Good point, fixed 👍